### PR TITLE
Feature/pt 2166 save custom field values when creating a subscription

### PR DESCRIPTION
**Pull request by i-just**
### Description
This PR makes saving custom field values against subscriptions much easier. 

​For example, you have a form that the user needs to fill out when purchasing a recurring product. That form has custom fields. The subscription element has a field layout with matching custom fields. When you submit the form, we should retain the custom field values and save them against the subscription when it is created following a successfully processed subscription creation event.

### Related issues
#25

### CODE REVIEW

File: src/controllers/CheckoutController.php [hunk 1]
```diff
@@ -53,15 +57,35 @@ public function actionCheckout(): Response
...
         }
 
+        $params = [];
+        $fields = $request->getBodyParam('fields');
+        if (!empty($fields)) {
+            // check the checkout mode - if it's subscription, proceed with creating a draft
+            $mode = $checkoutService->getCheckoutMode($lineItems);
+            if ($mode === StripeCheckoutSession::MODE_SUBSCRIPTION) {
+                // create an unpublished & unsaved draft subscription in Craft;
+                $subscription = new Subscription();
```
Comment (nfourtythree) [inline src/controllers/CheckoutController.php]: Worth changing this to use the `createObject` method, just for future-proofing and DI?

File: src/controllers/CheckoutController.php [hunk 2]
```diff
@@ -53,15 +57,35 @@ public function actionCheckout(): Response
...
+                $subscription->title = DateTimeHelper::now()->format('Y-m-d H:i:s');
+                $subscription->setFieldValuesFromRequest('fields');
+                if (Craft::$app->getDrafts()->saveElementAsDraft($subscription, markAsSaved: false)) {
+                    // send the uid of it to Stripe to be stored as metadata on the Session(!)
+                    $params['metadata']['uid'] = $subscription->uid;
```
Comment (nfourtythree) [inline src/controllers/CheckoutController.php]: As this metadata is stored in Stripe, could it be worth making this key a little less generic and more Craft specific. As an example we use `commerceTransactionHash` in Commerce just so it is obvious where the piece of data comes from.

Not a big deal just a suggestion (`craftSubscriptionUid` maybe?), I know it is a massive edge case but could also avoid anything else randomly changing that value

File: src/services/Subscriptions.php [hunk 3]
```diff
@@ -87,10 +87,13 @@ public function syncAllSubscriptions(): void
      * This takes the stripe subscription data from the API and creates or updates a Subscription element.
      *
      * @param StripeSubscription $subscription
+     * @param bool $checkForUnsavedDraft
      * @return bool Whether the synchronization succeeded.
      */
-    public function createOrUpdateSubscription(StripeSubscription $subscription): bool
+    public function createOrUpdateSubscription(StripeSubscription $subscription, bool $checkForUnsavedDraft = false): bool
```
Comment (nfourtythree) [inline src/services/Subscriptions.php]: Probably an edge case that anyone would be extending this class, but if they are and are re-implementing this method they would need to make sure their code is updated as the signature has changed.

File: src/services/Subscriptions.php [hunk 4]
```diff
@@ -99,8 +102,18 @@ public function createOrUpdateSubscription(StripeSubscription $subscription): bo
             ->one();
 
         if ($subscriptionElement === null) {
-            /** @var SubscriptionElement $subscriptionElement */
-            $subscriptionElement = new SubscriptionElement();
+            if ($checkForUnsavedDraft) {
```
Comment (nfourtythree) [inline src/services/Subscriptions.php]: Is this param needed? Could we not just check if the retrieved element is a draft and then do the logic later in this method based on that?

**Review by nfourtythree — CHANGES_REQUESTED** [review]

File: src/services/Subscriptions.php (hunk 4 above)
Comment (i-just) [reply src/services/Subscriptions.php]: To get the unsaved draft, we need to query Stripe to get the session by subscription id. From that we need to extract the `craftSubscriptionUid` metadata and use that to query for the unsaved draft (by uid and including drafts). 
So, for me, it makes more sense to have that param and do all that work only when needed, which should be only in response to receiving a `customer.subscription.created` webhook. 
I’m happy to discuss this further, though!

Comment (i-just) [reply src/services/Subscriptions.php]: It’s now updated as per our chat.

**Review by nfourtythree — APPROVED** [review]
Looks good to me!

**Pull request merged**